File: Data-Structures/Linked-Lists/reverse-a-linked-list.c
// Complete the reverse function below.

/*
 * For your reference:
 *
 * SinglyLinkedListNode {
 *     int data;
 *     SinglyLinkedListNode* next;
 * };
 *
 */
SinglyLinkedListNode* reverse(SinglyLinkedListNode* p) {
    if (p->next == NULL) {
        return p;
    }
    SinglyLinkedListNode* head = reverse(p->next);
    SinglyLinkedListNode* q = p->next;
    q->next = p;
    p->next = NULL;
    return head;
}

File: Data-Structures/Linked-Lists/reverse-a-linked-list.java
/*
  Reverse a linked list and return pointer to the head
  The input list will have at least one element  
  Node is defined as  
  class Node {
     int data;
     Node next;
  }
*/
    // This is a "method-only" submission. 
    // You only need to complete this method. 
Node Reverse(Node head) {
    Node p = null;
    Node q = head;
    Node r = null;
    while(q!=null) {
        r = q.next;
        q.next = p;
        p = q;
        q = r;
    }
    head = p;
    return head;
}
